src/data_structure/graph.rs

```diff
@@ -0,0 +1,294 @@
+    pub fn remove_node(&mut self, index: usize) -> Option<N> {
+        match self.nodes.remove(&index) {
+            Some(node) => {
+                // Only keep those edges that don't contain index as head or tail
+                self.edges
+                    .retain(|_, edge| edge.head != index && edge.tail != index);
+                return Some(node.weight);
+            }
+            None => {
+                return None;
+            }
+        }
+    }
+
+    /// Given two indices head and tail, attempt to add an edge whose
+    /// starting node has index head and whose ending node has index
+    /// tail to the graph. If this succeeds, return index assigned to
+    /// that edge. Otherwise return None.
+    pub fn add_edge(&mut self, weight: E, head: usize, tail: usize) -> Option<usize> {
+        if !self.nodes.contains_key(&head) || !self.nodes.contains_key(&head) {
+            None
+        } else {
+            let mut index = self.edges.len();
+
+            // There are at most self.edges.len() many indices in use. Hence
+            // trying self.edges.len()+1 many of them will always result
+            // in an available index. Most of the time, the first one
+            // (i.e. self.edges.len() ) will work but if we delete and
+            // insert a bunch of edges, this isn't necessarily true.
+            while self.edges.contains_key(&index) {
+                index -= 1;
+            }
+
+            self.edges.insert(
+                index,
+                Edge {
+                    weight: weight,
+                    head: head,
+                    tail: tail,
+                },
+            );
+            Some(index)
+        }
+    }
+
+    /// Remove edge indexed at 'index'
+    pub fn remove_edge(&mut self, index: usize) {
+        self.edges.remove(&index);
+    }
+
+    /// Return first index of an edge of the form head --> tail.  This
+    /// is not necessarily the unique index since we explicitly allow
+    /// multiple edges between two nodes.
+    pub fn find_edge(&self, head: usize, tail: usize) -> Option<usize> {
+        let edges: Vec<usize> = self.find_n_edges(1, head, tail);
+        if edges.len() == 0 {
+            return None;
+        } else {
+            return Some(edges[0]);
+        }
+    }
+
+    /// Return the indices of all edges of the form head --> tail
+    pub fn find_edges(&self, head: usize, tail: usize) -> Vec<usize> {
+        self.find_n_edges(0, head, tail)
+    }
+
+    /// Returns up to the first n edges of the form head --> tail. If
+    /// n == 0, it returns all available edges.
+    pub fn find_n_edges(&self, n: usize, head: usize, tail: usize) -> Vec<usize> {
+        let mut result: Vec<usize> = Vec::new();
+        for (index, edge) in self.edges.iter() {
+            if [edge.head, edge.tail] == [head, tail]
+                || (self.is_directed() && [edge.head, edge.tail] == [tail, head])
```

Comment:
I don't understand this line.

I think when our graph is *not* directed, edge from `tail` to `head` can be treated as edge from `head` to `tail`.

Reply:
Yea agree @plus7wist this looks little complicated.